python_interface/design/03_pyHorace_description.md

```diff
@@ -0,0 +1,269 @@
+The `py.*` namespace, nonetheless, can be used in the interface library for the Matlab interface, and an 
+[implementation](https://github.com/mducle/horace-euphonic-interface/blob/light_wrapper/%2Beuphonic/light_python_wrapper.m)
+suited to the needs of Euphonic and Brillem is currently in review.
+
+The `pyHorace` implementation, on the other hand will rely on `call_python` with Matlab-side and Python-side wrappers.
+On the Python side, a [`WrappedPythonClass`](https://github.com/mducle/hugo/blob/master/pyHorace/FunctionWrapper.py#L27)
+is used to wrap any Python class such that its objects are stored in a module-global dictionary and only a string reference is passed to Matlab.
+Additional functions, `get_obj_prop` and `call_obj_method` as well as wrapped constructors are defined
+and stored in a module-global dictionary for access by `call_python`.
+On the Matlab side, a [`pyclasswrapper`](https://github.com/mducle/hugo/blob/master/src/pyclasswrapper.m) 
+is used to wrap the these constructors and property access functions so as to present the same interface to the Matlab code as the pure Matlab case.
+The point of these indirections is that only string references rather than objects are passed between Matlab and Python
+but the behaviour is similar to creating objects and accessing their properties and methods directly.
+
+
+# Limitations and future work
+
+There are currently some limitations due to the architecture chosen for `pyHorace` which can be alleviated with more work:
+
+1. The Python syntax for any Matlab function/methods must exactly follow the Matlab syntax.
+2. Matlab functions can only be called as methods of the `Matlab` class and not independently (e.g. `without the m.* prefix`)
+3. Matlab help text cannot be accessed from Python.
+4. MPI routines don't work.
+
+Limitations 1 and 2 come from the choice, outlined in the [first section](#compiled-matlab) to use a gateway function which calls `feval`.
+This can be overcome by writing additional Python wrapper for each Matlab function or class we want to expose to Python.
+Alternatively the `Matlab` and `MatlabProxyObject` classes can be modified to automatically
+convert Python keyword arguments to `('name', value)` pairs to be passed to Matlab if the Matlab code is modified to accept them.
+
+Limitation 3 comes from a restriction of the compiled Matlab output itself:
+Matlab does not allow access to its online documentation system from compiled code.
+The only way to work around this is to save the help text for every Matlab function or class we want to expose to Python to a text file
+and then to either generate wrappers for each (such as that used to overcome limitations 1 and 2)
+or to modify the `Matlab` and `MatlabProxyObject` classes to look up these help text.
+
+Limitation 4 is because the MPI routines needs to launch additional instances of the Matlab interpreter which is not possible from within Python.
```

Comment:
This is a serious limitation(!). 

The creation of and execution of jobs through the MPI framework is a key element of the Horace performance improvements -- without parallelizing the algorithms there is only limited scope for performance improvements over what we currently have.

Reply:
As I mention below this should be possible to overcome. It would need changes to the `init` methods of `ClusterMPI` and `ClusterHerbert` (`ClusterParPool` should be fine I think). Essentially, instead of the code to launch a worker Matlab processes with `ProcessBuilder`, we would instead launch the compiled Matlab executables instead.